docs/adr/007_cups_firmware_upgrade.md

```diff
@@ -0,0 +1,129 @@
+contains the `package` field which indicates the current firmware version of the
+Basics Station. The value will need to be updated whenever a firmware upgrade of
+the Basics Station is performed by the user.
+
+### Firmware upgrade flow
+
+1. The `package` parameter, along with URL pointing to the location of the
+   firmware upgrade binary in a storage account, will be stored in the desired
+   properties of the device twin of the concentrator running the Basics Station.
+   Those values reflect the up-to-date firmware version that needs to be used
+   when communicating with the CUPS server. The firmware upgrade is initiated
+   through updating these parameters in the concentrator device twin.
+
+1. The Basics Station sends the CUPS request containing the currently used
+   values.
+
+1. The LNS determines if there are discrepanices between the values stored in
+   the concentrator twin and the ones provided by the Station. If there are
+   differences, the LNS will trigger the firmware upgrade by sending the correct
+   values to the Basics Station. The Station will then execute the actual
+   firmware upgrade.
+
+### IoT Hub related changes
+
+Only change is related to the concentrator device twin.
+
+In addition to the values already stored in the twin, the following need to be
+added:
+
+```json
+"cups": {
+  // ...
+  "package": "1.0.1",
+  "fwUrl": "https://...",
+  "fwKeyChecksum": 123456,
+  "fwSignature": ""
+}
+```
+
+- **'package'**: current package version of the Station
+- **'fwUrl'**: URL pointing to the storage location of the file required to run
+  the upgrade
+- **'fwKeyChecksum'**: checksum of the key used to sign the digest of the
+  firmware upgrade file
+- **'fwSignature'**: signature of the uploaded firmware upgrade file
+
+### Storage related changes
+
+We will use the same storage solution as the one selected for the general CUPS
+protocol support (storage account). A new container will be used to store the
+firmware upgrade files.
+
+### Azure Function related changes
+
+A new endpoint will be added in the Facade Azure Function which will be used to
+fetch firmware upgrade files from the storage account.
+
+### Changes in the LoRaWan Network Server
+
+The implementation of `CupsProtocolMessageProcessor` should be extended for
+checking the `package` field from the concentrator device twin. In case there
+the value is different from the one received from the Basics Station in the CUPS
+request, the Network Server will trigger the download of the firmware upgrade
+file (using the Facade Function) and populate the CUPS response accordingly, so
+that the Station can then execute the upgrade.
+
+### LoRa Device Provisioning CLI changes
+
+A new command will be added to the Device Provisioning CLI which will allow the
+user to trigger a firmware upgrade. The command will accept the follwing inputs:
+
+- firmware upgrade file
+- signature
+- CRC32 checksum of the key used for the signature
+- Basics Station package version
```

Comment:
```suggestion
- new package version
```

Reply:
i would omit "basics station package version" but just say new package version, since the update might decide to update not only the basic station but also other components